src/sim/detail/InitializeTracks.cu

```diff
@@ -0,0 +1,339 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file InitializeTracks.cu
+//---------------------------------------------------------------------------//
+#include "InitializeTracks.hh"
+
+#include <thrust/device_ptr.h>
+#include <thrust/reduce.h>
+#include <thrust/remove.h>
+#include <thrust/scan.h>
+#include <vector>
+#include "base/Atomics.hh"
+#include "base/DeviceVector.hh"
+#include "base/KernelParamCalculator.cuda.hh"
+
+namespace
+{
+using namespace celeritas;
+using celeritas::detail::flag_alive;
+//---------------------------------------------------------------------------//
+// HELPER CLASSES
+//---------------------------------------------------------------------------//
+struct IsEqual
+{
+    size_type value;
+
+    CELER_FUNCTION bool operator()(size_type x) const { return x == value; }
+};
+
+//---------------------------------------------------------------------------//
+// KERNELS
+//---------------------------------------------------------------------------//
+/*!
+ * Initialize the track states on device. The track initializers are created
+ * from either primary particles or secondaries. The new tracks are inserted
+ * into empty slots (vacancies) in the track vector.
+ */
+__global__ void init_tracks_kernel(const StatePointers            states,
+                                   const ParamPointers            params,
+                                   const TrackInitializerPointers inits)
+{
+    auto thread_id = KernelParamCalculator::thread_id().get();
+    if (thread_id < inits.vacancies.size())
+    {
+        // Get the track initializer, starting from the back of the vector
+        const TrackInitializer& init
+            = inits.initializers[inits.initializers.size() - thread_id - 1];
```

Comment:
Sure thing. Not exactly -- I think think the only case where we need to worry about the parent state being overwritten is when the parent track produces secondaries and is killed (if it doesn't produce secondaries no other track will need the state, and if it's alive nothing will be initialized there). This case is handled in the kernel that finds the vacancies and counts the secondaries: one of the secondaries is immediately copied into the dead track (preserving the geometry state) and the track is marked as alive. So when we get to this point, it should be safe to copy any of the states in any order.

Track initializers are pulled from the back of the vector because those are the most recently added (and therefore the ones that still might have a parent they can copy the geometry state from) and because there could be more initializers than empty track slots.

Reply:
Great! Thanks for the explanation.